src/NimBLERemoteCharacteristic.cpp

```diff
@@ -440,6 +441,23 @@ std::string NimBLERemoteCharacteristic::readValue() {
 } // readValue
 
 
+/**
+ * @brief Get the value of the remote characteristic.
+ * @return The value of the remote characteristic.
+ */
+std::string NimBLERemoteCharacteristic::getValue(time_t *timestamp) {
+    if(timestamp == nullptr) {
+        return m_value;
+    }
+
+    m_semaphoreReadCharEvt.take("readValue");
+    std::string value = m_value;
+    *timestamp = m_timestamp;
+    m_semaphoreReadCharEvt.give(1);
```

Comment:
I know it seems strange but 0 == success in BLE (also 14 in NimBLE) anything else is an error value. So this should be give(0) or just give().

Reply:
Actually this method should probably be refactored a bit since we always return a copy of `m_value` .
Something like:
```
{
    m_semaphoreReadCharEvt.take("readValue");
    std::string value = m_value;
    m_semaphoreReadCharEvt.give();

    if(timestamp != nullptr) {
        *timestamp = m_timestamp;
    }

    return value;
}
```

Reply:
I wrote it like that on purpose. If we change it to your suggestion there is a little chance that the return value and the timestamp do not belong together, if a notification is in the middle of an update.

I do not like the extra copy of the value but I could not come up with another way. Do you?

Reply:
Making a copy is necessary but the compiler should optimize it to copy directly into the callers container. 

Also now that you mention the timing issue, we have no protection in the notification handler against the value being updated as `getValue()` is copying. The semaphore would only block if `readValue()` was called at the same time, but that would be odd.

Of course we cannot take a semaphore in the notification handler unless it was used with a 0 timeout then tested for success, but then we wouldn't be able to update the value if it failed.

Reply:
I am a semaphore zero  and you a h(2z)ero+e, what would you suggest?

Reply:
Considering the short term nature I think we could block in the notify event for a short time but it would require the use of it's own (new) semaphore to ensure that it would only block for the `getValue()` copy operation. 

So we could create `m_semaphoreNotifyEvt` then put `m_semaphoreNotifyEvt.take()` and `m_semaphoreNotify.give()` where it currently is in `getValue()` and then do the same [here](https://github.com/h2zero/NimBLE-Arduino/pull/49/files#diff-8f692c893f6a420d294d6426dcac9a9cR722). 

Would require stress testing but I don't think there will be any issues. The downside to this is an extra 200 bytes of memory per characteristic.

Reply:
Could you do it? I am really no good in semaphores...

Reply:
The 200 bytes is quite a lot, but I think it is worth it to have an characteristic value that is updated by notification. Will your dynamic semaphore trick make it any  better?

Reply:
I'll try it locally, if it works well I'll push the changes to this. 

Dynamic semaphore creation is a possibility here at the cost of speed for setup/alloc/dealloc, same as the other ones, but I think those can be replaced with task notifications.

Reply:
Thnx!

Reply:
I put this through a lot of stress testing for many hours. I used a loop calling `getValue()` every random 8-12us and a peripheral sending a notification every 50ms.

*  With no semaphore use at all there was no segfaults (wouldn't want to trust that, very odd).

* With static semaphore, no problem and appears safe.

* With dynamic semaphore, massive memory leaks due slow alloc/dealloc and not enough time for it to release the memory. Also a fault in the `take()` method as it took the semaphore while it was being deleted (at least this exposed a flaw in that idea and needs new thinking). 

Now for real world use I would consider the static semaphore route for the time being to stay on the safe side, if you agree I will push that addition to this PR. I hope for dynamic or another alternative to work in the future as the semaphores are expensive.

Reply:
You've put a lot of work in this, thank you so much for that!

> calling getValue() every random 8-12us and a peripheral sending a notification every 50ms

You are really pushing it to the limits!!

> With no semaphore use at all there was no segfaults (wouldn't want to trust that, very odd)

Can we come up with a reason for this? My conclusion was equal, although I did not call `getValue()` that frequently. Again, I am no semepahore expert, nor have I any knowledge of the ESP32 OS, but: could it be that something in FreeRTOS and/or the way that the running of programs and core functionality is organized on one or two cores prevents the occurrence of a segfault? Could it be that the semaphores are only necessary if the main programmer spawns more than one thread running BLE functions? 

> I would consider the static semaphore route for the time being to stay on the safe side

I agree with you, if we can't explain why segfaults aren't there, it is best to stay at the save side.

>  if you agree I will push that addition to this PR

Of course! I'd be very happy with that!

Reply:
>Can we come up with a reason for this?

My first thought is when we change the value in notify it creates a new string object and just changes the location pointer of m_value. If the copy constructor was called in `getValue()` at the same time it would still get a valid reference, at least until the original m_value destructor is called, which could be long enough to finish the copy. I could be completely wrong, worth testing though. 

The other possibility is flawed testing 😄, I just realized the peripheral did not vary the data length so my test never went out of range.

Reply:
> If the copy constructor was called in getValue() at the same time it would still get a valid reference, at least until the original m_value destructor is called, which could be long enough to finish the copy

I do not know how FreeRTOS switches tasks. We're not interrupt based, interrupts may occur everywhere. But maybe the OS has some mechanism to protect for the segfaults? idk...

Reply:
As far as I know it does not have any such mechanism. It simply switches between tasks on a priority basis, when one task is done or waiting for an event it switches to the next highest priority task. Interrupts are handled as their own, highest priority, task that stops all others I believe, but it's been a while since I got into that.

Reply:
I have come up with a compromise for this, I'll push it to this PR right away, stress testing for a hour now without a problem. It uses the same semaphore with a 0 timeout in the notify callback, that does not update the value if it fails to take it, but that should only happen when reading directly at the same time, so it should not be an issue 99.99% of the time. 

If you could test it later on your setup and let me know I'd really appreciate it.

Reply:
> If you could test it later on your setup and let me know I'd really appreciate it.

Sure! But most likely not today because I am busy at the office. Tomorrow for sure!